File: src/core/shell.js
import { spawn } from "node:child_process";

export async function runCommand(command, args = [], options = {}) {
  const {
    cwd,
    env,
    timeoutMs = 120_000,
    input,
    allowFailure = false
  } = options;

  return await new Promise((resolve, reject) => {
    const startedAt = Date.now();
    const child = spawn(command, args, {
      cwd,
      env: { ...process.env, ...env },
      shell: false,
      stdio: ["pipe", "pipe", "pipe"]
    });

    let stdout = "";
    let stderr = "";
    const timer = setTimeout(() => {
      child.kill("SIGTERM");
    }, timeoutMs);

    child.stdout.on("data", (chunk) => {
      stdout += chunk.toString();
    });
    child.stderr.on("data", (chunk) => {
      stderr += chunk.toString();
    });
    child.on("error", (error) => {
      clearTimeout(timer);
      reject(error);
    });
    child.on("close", (code, signal) => {
      clearTimeout(timer);
      const result = {
        command: [command, ...args].join(" "),
        cwd,
        code,
        signal,
        stdout,
        stderr,
        durationMs: Date.now() - startedAt,
        ok: code === 0
      };
      if (!result.ok && !allowFailure) {
        const detail = stderr.trim() || stdout.trim() || `exit code ${code}`;
        reject(new Error(`${result.command} failed: ${detail}`));
      } else {
        resolve(result);
      }
    });

    if (input) {
      child.stdin.write(input);
    }
    child.stdin.end();
  });
}

export async function commandExists(command) {
  const result = await runCommand("command", ["-v", command], {
    allowFailure: true,
    timeoutMs: 5_000
  }).catch(() => ({ ok: false }));
  return result.ok;
}

export function splitShellCommand(command) {
  const tokens = [];
  let current = "";
  let quote = null;
  for (let index = 0; index < command.length; index += 1) {
    const char = command[index];
    if (quote) {
      if (char === quote) {
        quote = null;
      } else {
        current += char;
      }
    } else if (char === "'" || char === "\"") {
      quote = char;
    } else if (/\s/.test(char)) {
      if (current) {
        tokens.push(current);
        current = "";
      }
    } else {
      current += char;
    }
  }
  if (current) {
    tokens.push(current);
  }
  return tokens;
}

File: go_issue_agent/core/shell.py
from __future__ import annotations

import shlex
import subprocess
import time


def run_command(command: str, args: list[str] | None = None, cwd=None, env=None, timeout_ms: int = 120000, input_text: str | None = None, allow_failure: bool = False):
    args = args or []
    started_at = time.time()
    merged_env = None
    if env:
        import os

        merged_env = dict(os.environ)
        merged_env.update(env)

    try:
        completed = subprocess.run(
            [command, *args],
            cwd=cwd,
            env=merged_env,
            input=input_text,
            text=True,
            capture_output=True,
            timeout=timeout_ms / 1000,
            check=False,
        )
    except subprocess.TimeoutExpired as error:
        result = {
            "command": " ".join([command, *args]),
            "cwd": cwd,
            "code": None,
            "signal": "timeout",
            "stdout": error.stdout or "",
            "stderr": error.stderr or "",
            "durationMs": int((time.time() - started_at) * 1000),
            "ok": False,
        }
        if allow_failure:
            return result
        raise RuntimeError(f'{result["command"]} failed: timeout')

    result = {
        "command": " ".join([command, *args]),
        "cwd": cwd,
        "code": completed.returncode,
        "signal": None,
        "stdout": completed.stdout,
        "stderr": completed.stderr,
        "durationMs": int((time.time() - started_at) * 1000),
        "ok": completed.returncode == 0,
    }
    if not result["ok"] and not allow_failure:
        detail = completed.stderr.strip() or completed.stdout.strip() or f'exit code {completed.returncode}'
        raise RuntimeError(f'{result["command"]} failed: {detail}')
    return result


def split_shell_command(command: str) -> list[str]:
    return shlex.split(command)
